src/data/set/ncard.lean

```diff
@@ -0,0 +1,696 @@
+
+lemma card_insert_le (a : α) (s : set α) : (insert a s).ncard ≤ s.ncard + 1 :=
+begin
+  obtain (hs | hs) := s.finite_or_infinite,
+  { exact (em (a ∈ s)).elim (λ h, (ncard_insert_of_mem h).trans_le (nat.le_succ _))
+      (λ h, by rw ncard_insert_of_not_mem h hs)},
+  rw (hs.mono (subset_insert a s)).ncard,
+  exact nat.zero_le _,
+end
+
+lemma ncard_insert_eq_ite (hs : s.finite . to_finite_tac) :
+  ncard (insert a s) = if a ∈ s then s.ncard else s.ncard + 1 :=
+begin
+  by_cases h : a ∈ s,
+  { rw [ncard_insert_of_mem h, if_pos h] },
+  { rw [ncard_insert_of_not_mem h hs, if_neg h] }
+end
+
+@[simp] lemma card_doubleton (h : a ≠ b) : ({a, b} : set α).ncard = 2 :=
+by {rw [ncard_insert_of_not_mem, ncard_singleton], simpa}
+
+@[simp] lemma ncard_diff_singleton_add_one (h : a ∈ s) (hs : s.finite . to_finite_tac) :
+  (s \ {a}).ncard + 1 = s.ncard :=
+begin
+  have h' : a ∉ s \ {a}, by {rw [mem_diff_singleton], tauto},
+  rw ←ncard_insert_of_not_mem h' (hs.diff {a}),
+  congr',
+  simpa,
+end
+
+@[simp] lemma ncard_diff_singleton_of_mem (h : a ∈ s) (hs : s.finite . to_finite_tac) :
+  (s \ {a}).ncard = s.ncard - 1 :=
+eq_tsub_of_add_eq (ncard_diff_singleton_add_one h hs)
+
+lemma ncard_diff_singleton_lt_of_mem (h : a ∈ s) (hs : s.finite . to_finite_tac) :
+  (s \ {a}).ncard < s.ncard :=
+by {rw [←ncard_diff_singleton_add_one h hs], apply lt_add_one}
+
+lemma ncard_diff_singleton_le (s : set α) (a : α) :
+  (s \ {a}).ncard ≤ s.ncard :=
+begin
+  obtain (hs | hs) := s.finite_or_infinite,
+  { apply ncard_le_of_subset (diff_subset _ _) hs},
+  convert zero_le _,
+  exact (hs.diff (by simp : set.finite {a})).ncard,
+end
+
+lemma pred_ncard_le_ncard_diff_singleton (s : set α) (a : α) :
+  s.ncard - 1 ≤ (s \ {a}).ncard :=
+begin
+  cases s.finite_or_infinite with hs hs,
+  { by_cases h : a ∈ s,
+    { rw ncard_diff_singleton_of_mem h hs, },
+    rw diff_singleton_eq_self h,
+    apply nat.pred_le},
+  convert nat.zero_le _,
+  rw hs.ncard,
+end
+
+lemma ncard_exchange (ha : a ∉ s) (hb : b ∈ s) :
+  (insert a (s \ {b})).ncard = s.ncard :=
+begin
+  cases s.finite_or_infinite with h h,
+  { haveI := h.to_subtype,
+    rw [ncard_insert_of_not_mem, ncard_diff_singleton_add_one hb],
+    simpa only [mem_diff, not_and] using ha},
+  rw [((h.diff (set.to_finite {b})).mono (subset_insert _ _)).ncard, h.ncard],
+end
+
+lemma ncard_exchange' (ha : a ∉ s) (hb : b ∈ s) :
+  ((insert a s) \ {b}).ncard = s.ncard :=
+by rw [←ncard_exchange ha hb, ←singleton_union, ←singleton_union, union_diff_distrib,
+    @diff_singleton_eq_self _ b {a} (λ h, ha (by rwa ← mem_singleton_iff.mp h) )]
+
+end insert_erase
+
+lemma ncard_image_le (hs : s.finite . to_finite_tac) :
+  (f '' s).ncard ≤ s.ncard :=
+begin
+  rw ncard_eq_to_finset_card s hs,
+  haveI := hs.fintype,
+  convert @finset.card_image_le _ _ s.to_finset f _,
+  rw [ncard_eq_to_finset_card, finite.to_finset_image _ hs],
+  { congr', rw [←finset.coe_inj, finite.coe_to_finset, coe_to_finset]},
+  { apply_instance},
```

Comment:
There's a lot of spacing fixes like this to be done throughout the code.
```suggestion
  { congr', rw [←finset.coe_inj, finite.coe_to_finset, coe_to_finset] },
  { apply_instance },
```

Reply:
Thanks for the review! I think I've addressed your comments.